text/0000-alloc-me-maybe.md

```diff
@@ -0,0 +1,547 @@
+
+Unfortunately, attempting to handle allocation failure in C(++) has been a historical source of critical vulnerabilities. For instance, if reallocating an array fails but isn't noticed, the user of the array can end up thinking it has more space than it actually does and writing past the end of the allocation.
+
+The return-value-based approach is problematic because neither language has good facilities for mandating that a result is actually *checked*. There are two notable cases here: when the result of the allocation is some kind of error code (e.g. a bool), or the result is a pointer into the allocation (or a specific pointer indicating failure).
+
+In the error code case, neither language provides a native facility to mandate that error codes must be checked. However compiler-specific attributes like GCC's warn_unused_result can be used here. Unfortunately nothing mandates that the error code is used *correctly*. In the pointer case, blindly dereferencing is considered a valid use, fooling basic lints.
+
+Unwinding is better than error codes in this regard, because completely ignoring an exception aborts the process. The author's understanding is that problems arise from the complicated exception-safety rules C++ collections have.
+
+Both of these concerns are partially mitigated in Rust. For return values, Result and bool have proper on-by-default must-use checks. However again nothing mandates they are used properly. In the pointer case, we can however prevent you from ever getting the pointer if the Result is an `Err`. For unwinding, it's much harder to run afoul of exception-safety in Rust, especially since copy/move can't be overloaded. However unsafe code may have trouble.
+
+
+
+
+## Additional Background: Overcommit and Killers
+
+Some operating systems can be configured to pretend there's more memory than there actually is. Generally this is the result of pretending to allocate physical pages of memory, but only actually doing so when the page is accessed. For instance, forking a process is supposed to create two separate copies of the process's memory, but this can be avoided by simply marking all the pages as *copy on write* and having the processes share the same physical memory. The first process to mutate the shared page triggers a page fault, which the OS handles by properly allocating a new physical page for it. Similarly, to postpone zeroing fresh pages of memory, the OS may use a copy-on-write zero page.
+
+The result of this is that allocation failure may happen when memory is first *accessed* and not when it's actually requested. If this happens, someone needs to give up their memory, which can mean the OS killing your process (or another random one!).
+
+This strategy is used on many *nix variants/descendants, including Android, iOS, MacOS, and Ubuntu.
```

Comment:
Nit: macOS doesn't have Linux-like OOM killer. On OOM it may ask user to choose an app to kill, but that seems to be a high-level feature. The system itself just keeps eating swap space until the machine becomes unusable.

Reply:
Don't know about other linuxes, but Ubuntu doesn't have one either (or a broken one, as machine becomes unusable anyway)

Reply:
All (somewhat modern) Linuxes including Ubuntu have a in kernel OOM killer.

*TL;DR:* Linux OOM Killer is for many use-cases kinda broken, there are ways to work around it, things are slowly changing.  Rest of the post is slightly out of context.

---

But it's a last resort fallback which only triggers if there is really no other choice. So your system is likely to (seem to) freeze before it triggers. 

Reasons for this include:

- swap being slow (but don't disable it, doing so easily can make it worse)
- high memory pressure making your system slow even before it runs out of memory
- OOM killer only triggering if there *really* is no other choice and itself not being the fastest

While you can fine tune it in many ways non really yield a satisfying result for desktop systems in my experience as it is fundamentally the last fallback which only triggers in the worst situation.

But not only are there kernel patches with alternative OOM Killers,  OOM killing can also be instrumented from user-space. With this programs like [earlyroom](https://github.com/rfjakob/earlyoom) (I haven't tried it out!) can be used to do OOM killing once memory pressure becomes high, killing memory hugging programs before your system becomes so slow that it seem to be frozen. A mac like OOM killer experience should be possible to be created in this way, too.

Lastly there is systemd-oomd a systemd service which like earlyroom is a user space OOM Killer and which is more focused on a per-slice/service killing, combined with the movement to launch all applications through the systemd user manager (default on e.g. GNOME) it might be the default "better"  OOM Killer in the future which current desktop Linuxes miss ("better" ~= it's not actually better but just has different goals).

One think which is important to be aware of is that **even with over-commit allocations can still fail** especially on large allocations the reasons for this can vary and I'm sure this is discussed by enough other commits in this RFC.  But this is the reason why I like the `try_{reserve,*}` methods as they do a good enough job for many use cases, potentially even with memory over-commit enabled.